File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3093_longest-common-suffix-queries/solutions/cpp/solution.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> stringIndices(vector<string>& wordsContainer, vector<string>& wordsQuery) {
        vector<int> result;
        
        for (const string& query : wordsQuery) {
            int maxLength = -1;
            int minIndex = -1;
            int minLength = numeric_limits<int>::max();

            for (int i = 0; i < wordsContainer.size(); ++i) {
                const string& containerWord = wordsContainer[i];
                int commonLength = 0;
                int l = containerWord.length() - 1;
                int r = query.length() - 1;
                
                // Calculate longest common suffix length
                while (l >= 0 && r >= 0 && containerWord[l] == query[r]) {
                    ++commonLength;
                    --l;
                    --r;
                }
                
                if (commonLength > maxLength || (commonLength == maxLength && containerWord.length() < minLength) || (commonLength == maxLength && containerWord.length() == minLength && (minIndex == -1 || i < minIndex))) {
                    maxLength = commonLength;
                    minIndex = i;
                    minLength = containerWord.length();
                }
            }
            
            result.push_back(minIndex);
        }
        
        return result;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3093_longest-common-suffix-queries/solutions/java/solution.java
import java.util.Arrays;

public class Solution {
    public int[] stringIndices(String[] wordsContainer, String[] wordsQuery) {
        int[] ans = new int[wordsQuery.length];
        Arrays.fill(ans, -1);

        for (int i = 0; i < wordsQuery.length; i++) {
            String query = wordsQuery[i];
            int maxLength = -1;
            int minIndex = -1;
            int minLength = Integer.MAX_VALUE;

            for (int j = 0; j < wordsContainer.length; j++) {
                String container = wordsContainer[j];
                int commonLength = commonSuffixLength(container, query);

                if (commonLength > maxLength ||
                    (commonLength == maxLength && container.length() < minLength) ||
                    (commonLength == maxLength && container.length() == minLength && j < minIndex)) {
                    maxLength = commonLength;
                    minIndex = j;
                    minLength = container.length();
                }
            }
            ans[i] = minIndex;
        }
        return ans;
    }

    private int commonSuffixLength(String a, String b) {
        int length = 0;
        int aIndex = a.length() - 1;
        int bIndex = b.length() - 1;
        while (aIndex >= 0 && bIndex >= 0 && a.charAt(aIndex) == b.charAt(bIndex)) {
            length++;
            aIndex--;
            bIndex--;
        }
        return length;
    }
}
